### PR TITLE
Ascii code

### LINKED ISSUE #38: ASCII_Code

### PR DESCRIPTION
**Pull request by Zeinab15**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/ascii_code.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        raise AssertionError("Input must be a string")
+    # Check if the input is an empty string
+    if text == "":
+        return []
+    # Convert the string to a list of ASCII codes
+    c = []
+    # Iterate over the characters in the input string
+    for i in text:
+        c.append(ord(i))
+    return c
```
Comment (Khusro-S) [inline solutions/ascii_code.py:51]: The variable name `c` doesn't clearly explain what it's for. It would be better to rename it to something like `ascii_values` or `unicode_codes`. This will make the code easier to understand, as the name will show that the list is holding the `ASCII` or `Unicode` values of the characters in the string.

File: solutions/ascii_code.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        e.g. ascii_code(123)
+    AssertionError: if the input is an empty string (no characters) e.g. ascii_code("")
+    and also if the input is not a string eg. ascii_code(123) or ascii_code(12.3).
+
+    Example:
+    ascii_code('abc') -> [97, 98, 99]
+    ascii_code('!') -> [33]
+    ascii_code('') -> []
+    ascii_code(' ') -> [32]
+    ascii_code('0') -> [48]
```
Comment (Khusro-S) [inline solutions/ascii_code.py:37]:
```suggestion
    Example:
    >>> ascii_code('abc')
    [97, 98, 99]
    
    >>> ascii_code('!')
    [33]
    
    >>> ascii_code('')
    []
    
    >>> ascii_code(' ')
    [32]
    
    >>> ascii_code('0')
    [48]
```

Comment (Khusro-S) [reply solutions/ascii_code.py:37]: This change will allow you to test the doctests. Previously, it was just a docstring without the ability to run tests.

**Review by Khusro-S — CHANGES_REQUESTED** [review]
Overall, the code is pretty self-explanatory, clean, and nicely written. I just suggested a few small changes that I think would make things even clearer or more helpful 🎉

File: solutions/ascii_code.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    text: a string of characters
+
+    Returns:
+    a list of integers, where each integer is the ASCII code of the character
+    at the corresponding index in the input string
+
+    Raises:
+    TypeError: if the input is not a string
+        e.g. ascii_code(123)
+    AssertionError: if the input is an empty string (no characters) e.g. ascii_code("")
```
Comment (GaiSamuel) [inline solutions/ascii_code.py:29]: ascii_code("") wouldn't raise an assertion error since you modified it in the function in line 45 and for the rest, like ascii_code(123) or ascii_code(12.3) those are invalid inputs and wouldn't raise assertion error but type error. So line 29 and 30 need to be deleted.


Screenshot (67)

File: solutions/tests/test_ascii_code.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    def test_whitespace(self):
+        """Test conversion of whitespace characters."""
+        self.assertEqual(ascii_code(" "), [32])  # space
+
+    def test_empty_input(self):
+        """Test handling of empty string input."""
+        self.assertEqual(ascii_code(""), [])
+
+    def test_number_input(self):
+        """Test handling of number input."""
```
Comment (GaiSamuel) [inline solutions/tests/test_ascii_code.py:60]: The docstring needs to be changed to testing invalid input since the test is an invalid input. The rest of the tests from 64-74 are invalid inputs as well and return type error instead. So it would be better if they were all under one test.

Screenshot (65)

**Review by GaiSamuel — CHANGES_REQUESTED** [review]
Most of the code looks good however, some arrears would need changes and i have suggested a few.

File: solutions/ascii_code.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    ascii_code('abc') -> [97, 98, 99]
+    ascii_code('!') -> [33]
+    ascii_code('') -> []
+    ascii_code(' ') -> [32]
+    ascii_code('0') -> [48]
+
+    """
+    # Check if the input is not a string
+    if not isinstance(text, str):
+        raise AssertionError("Input must be a string")
```
Comment (GaiSamuel) [inline solutions/ascii_code.py:42]: would raise type error not assertion error.

**Review by GaiSamuel — APPROVED** [review]

File: solutions/tests/test_ascii_code.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    def test_whitespace(self):
+        """Test conversion of whitespace characters."""
+        self.assertEqual(ascii_code(" "), [32])  # space
+
+    def test_empty_input(self):
+        """Test handling of empty string input."""
+        self.assertEqual(ascii_code(""), [])
+
+    def test_None_input(self):
```
Comment (Khusro-S) [inline solutions/tests/test_ascii_code.py:59]: To keep the naming style consistent, I noticed that `test_None_input` doesn’t fully follow the snake_case convention, all the letters should ideally be lowercase, like `test_none_input`. Not a big deal, but I thought it’d be worth mentioning 😅

**Review by Khusro-S — APPROVED** [review]
